### PR TITLE
feat(web): Add web server command to serve JSON resume from URL with periodic refresh

### PR DESCRIPTION
**Pull request by kiraum**
This PR introduces a new `web` command for `serve` allowing to serving an JSON resume from a URL with periodic refresh. Key features include:

- New `web` command in `__main__.py` with options for URL, refresh interval, port, and host;
- Implementation of `WebHandler` class in `server.py` to handle fetching, rendering, and serving the resume;
- Periodic refresh of resume data from the specified URL;
- Caching mechanism to serve the last valid render if fetching fails.

This feature enables users to serve their resume from external sources, making it easier to keep the displayed resume up-to-date without manual intervention.

Notes:
- I did run all tests manually required on devbox, but I don't use devbox/nix (I try to install as less software as possible on my laptop);
-- On my setup, I normally include `optional-dependencies` on `pyproject`, with that I can use `uv pip compile --extra dev > requirements-dev.txt`, and have a local pre-commit.
- Built the image and seems to be working as expected, it's on production here: `curl https://ansiv.xpto.it`;
- Very nice change moving from `poetry` to `uv`;

Please, let me know if something doesn't look good, and/or if I missed something, and feel free to reject/close the PR if it's not something aligned with the project.

Thanks for the very nice project/repo/code, learning a lot with it! =D

### CODE REVIEW

File: ancv/web/server.py [hunk 1]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        Args:
+            request: The incoming web request.
+
+        Returns:
+            The response containing the rendered resume.
+        """
+        log = LOGGER.bind(request=request)
+        session: ClientSession = request.app["client_session"]
+
+        current_time = time.time()
```
Comment (alexpovel) [inline ancv/web/server.py:369]: Similar to timedelta above, let's have this...

- either be a stronger time: https://docs.python.org/3/library/datetime.html#datetime.datetime.now
- or, maybe "semantically better", a monotonic tick: https://docs.python.org/3/library/time.html#time.monotonic

`time.time()` has problems like:

> it can return a lower value than a previous call if the system clock has been set back between the two calls.

which probably messes with the cache.

File: ancv/web/server.py [hunk 2]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        log.debug("Started client session.")
+
+        log.debug("App context initialization done, yielding.")
+        yield
+
+        log.debug("App context teardown starting.")
+        await session.close()
+        log.debug("App context teardown done.")
+
+    async def fetch_and_render(self, session: ClientSession) -> str:
```
Comment (alexpovel) [inline ancv/web/server.py:317]: It's probably a good idea to split `fetch` and `render` up. `render` is a pure function IIRC, and `fetch` will do I/O.

You can try and introduce a new top-level function which calls three separate ones: fetch, render, cache. Not totally sure that will be an improvement or even necessary, but splitting `fetch_and_render` is probably a good idea! We can experiment with this. It will make testing much simpler probably.

File: ancv/web/server.py [hunk 3]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+            try:
+                self.cache = await self.fetch_and_render(session)
+                self.last_fetch = current_time
+            except Exception as e:
+                log.error("Failed to fetch resume", error=str(e))
+                if self.cache is None:
+                    raise
+
+        log.debug("Serving rendered template.")
+        return web.Response(text=self.cache)
```
Comment (alexpovel) [inline ancv/web/server.py:387]: I think this won't be able to do type narrowing properly? I suspect `self.cache` is detected as `Optional[str]` here (do you have an editor which lets you check this?)?

If yes, let's move this into the `else` block of `if should_refresh`. That could/should allow `mypy` etc. to detect that `self.cache` is `str`, as `is None` was excluded at that point.

File: ancv/web/server.py [hunk 4]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+    def __init__(self, destination: str, refresh_interval: int = 300) -> None:
+        """Initializes the handler.
+
+        Args:
+            destination: The URL to load the JSON Resume from.
+            refresh_interval: How often to refresh the resume in seconds.
+        """
+        self.destination = destination
+        self.refresh_interval = refresh_interval
+        self.cache: Optional[str] = None
```
Comment (alexpovel) [inline ancv/web/server.py:281]: This could be `self.cache: Optional[ResumeSchema] = None`, see below for more!

File: ancv/web/server.py [hunk 5]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        try:
+            async with session.get(self.destination) as response:
+                if response.status != 200:
+                    raise web.HTTPBadRequest(
+                        reason=f"Failed to fetch resume from {self.destination}"
+                    )
+                content = await response.text()
+                resume_json = json.loads(content)
+
+            with tempfile.NamedTemporaryFile(mode="w", suffix=".json") as tmp:
```
[the remark this answers is not in the document]
Comment (alexpovel) [reply ancv/web/server.py:339]: On second thought, this cache here is also busted when the server restarts, as there's no way to retrieve the file path of `NamedTemporaryFile` once it's lost?

If you only created a temp file to create a `Template`, can you try with

https://github.com/alexpovel/ancv/blob/99d6a1807b23112f21bc82fc99a5129962ab95db/ancv/visualization/templates.py#L204

?

**Review by alexpovel — COMMENTED** [review]
Thanks for your PR! Looks nice! 🙂 

A couple notes:

- devbox is totally optional, so if you can replicate it locally more or less, that's totally fine. CI uses devbox though, and that ultimately needs to pass. That would be the strongest argument in favor of devbox. That said, I've been thinking of ditching devbox too, because...
- ...I only used it for `poetry`, which sure enough broke down immediately when I tried to get ancv running locally again. `uv` seems much more robust by itself!
- does `[tool.uv] dev-dependencies` in `pyproject.toml` work for you? It has all dev dependencies there. I'm new to `uv` so happy to change that if there's a better way

CI is currently failing, I think because of GitHub-specific API token problems... so never mind that for now.

Two general notes:

- you can add your new functionality to docs, e.g. the _Selfhosting_ section of the README.
- a test will be needed I'm afraid! It should check for basic functionality. You you try and set the `destination` to a local server like https://docs.python.org/3/library/http.server.html , which is spun up for the test (similar to https://pkg.go.dev/net/http/httptest#NewServer if you're more familiar with Go). Checking the caching functionality would also be nice, you could set `refresh` to a second and play with that

**Comment by kiraum** [comment]
> Thanks for your PR! Looks nice! 🙂
> 
> A couple notes:
> 
>     * devbox is totally optional, so if you can replicate it locally more or less, that's totally fine. CI uses devbox though, and that ultimately needs to pass. That would be the strongest argument in favor of devbox. That said, I've been thinking of ditching devbox too, because...
> 
>     * ...I only used it for `poetry`, which sure enough broke down immediately when I tried to get ancv running locally again. `uv` seems much more robust by itself!
> 
>     * does `[tool.uv] dev-dependencies` in `pyproject.toml` work for you? It has all dev dependencies there. I'm new to `uv` so happy to change that if there's a better way
> 
> 
> CI is currently failing, I think because of GitHub-specific API token problems... so never mind that for now.
> 
> Two general notes:
> 
>     * you can add your new functionality to docs, e.g. the _Selfhosting_ section of the README.
> 
>     * a test will be needed I'm afraid! It should check for basic functionality. You you try and set the `destination` to a local server like https://docs.python.org/3/library/http.server.html , which is spun up for the test (similar to https://pkg.go.dev/net/http/httptest#NewServer if you're more familiar with Go). Checking the caching functionality would also be nice, you could set `refresh` to a second and play with that


Thanks for the quick answer/comments, I will go over it during the week, a bit slow here, because it's herfstvakantie (autumn holidays), so a bit busy with family! 😅 

I never tested `[tool.uv] dev-dependencies`, but I will take a look, I'm also new to `uv` and `ruff`, but really happy with it!

I normally use something like this:
```
» git diff pyproject.toml
diff --git a/pyproject.toml b/pyproject.toml
index 7ded164..9af67d5 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -41,6 +41,13 @@ dependencies = [
     "pydantic[email]>=2.9.2",
 ]

+[project.optional-dependencies]
+dev = [
+    "black>=23.0",
+    "mypy>=1.0",
+    "ruff>=0.0.290",
+]
+
 [project.urls]
 Homepage = "https://ancv.io"
 Repository = "https://github.com/alexpovel/ancv/"
```

With that, I can compile the requirements and requiments only for dev env:
````
» uv pip compile pyproject.toml > requirements.txt
Resolved 32 packages in 284ms
» uv pip compile pyproject.toml --extra dev > requirements-dev.txt
Resolved 39 packages in 191ms
````

Check the final diff:
````
» diff -U0 requirements.txt requirements-dev.txt | grep -vE '^@|#'
--- requirements.txt    2024-10-27 18:47:56
+++ requirements-dev.txt        2024-10-27 18:48:15
+black==24.10.0
+mypy==1.13.0
+mypy-extensions==1.0.0
+packaging==24.1
+pathspec==0.12.1
+platformdirs==4.3.6
+ruff==0.7.1
````

I will include the docs for the new functionality and tests, I'm trying to keep style as similar as I can, feel free to nitpick.
For the comments, we can go case by case, I will adding comments when pushing code, so you can check and resolve if good! =]

File: ancv/web/server.py [hunk 6]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
         f"{name.replace(' ', '-')};dur={duration // timedelta(milliseconds=1)}"
         for name, duration in timings.items()
     )
+
+
+class WebHandler(Runnable):
+    """A handler serving a rendered template loaded from a URL with periodic refresh."""
+
+    def __init__(self, destination: str, refresh_interval: int = 300) -> None:
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done. I'm just not using `int()`, because Typer already handles the type conversion since we specified refresh: int in the parameter type annotation, do you see the need to do it differently?

File: ancv/web/server.py [hunk 7]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        Returns:
+            The rendered template as string.
+        """
+        log = get_logger()
+        stopwatch = Stopwatch()
+        stopwatch(segment="Fetch Resume")
+
+        try:
+            async with session.get(self.destination) as response:
+                if response.status != 200:
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done. Out of curiosity, why do you prefer name convention? Readability or any other reason?

File: ancv/web/server.py [hunk 8]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+                    raise web.HTTPBadRequest(
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done. Moved to return to keep it the same as the other handlers. The main diff that I see is that `return` give us more direct control over the response, and raise immediately stop, which may not be the required case here.

File: ancv/web/server.py [hunk 9]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+                tmp.flush()
+                template = Template.from_file(Path(tmp.name))
+                rendered = template.render()
+
+            log.info("Successfully refreshed resume data")
+            self._last_valid_render = rendered
+            return rendered
+
+        except Exception as e:
+            if hasattr(self, "_last_valid_render"):
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done.

File: ancv/web/server.py (hunk 1 above)
Comment (kiraum) [reply ancv/web/server.py]: Done, and nice catch, moved to `time.monotonic()`. TIL

File: ancv/web/server.py (hunk 2 above)
Comment (kiraum) [reply ancv/web/server.py]: I like the idea to split, makes much more sense and keep it cleaner and easier to maintain, done.

File: ancv/web/server.py [hunk 10]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+                json.dump(resume_json, tmp)
+                tmp.flush()
+                template = Template.from_file(Path(tmp.name))
+                rendered = template.render()
+
+            log.info("Successfully refreshed resume data")
+            self._last_valid_render = rendered
+            return rendered
+
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done when splitting the function.

File: ancv/web/server.py (hunk 3 above)
Comment (kiraum) [reply ancv/web/server.py:421]: Changed the logic when doing the function split, so now it's a `str`, not `Optional` anymore. =]

File: ancv/web/server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexpovel) [reply ancv/web/server.py]: Nope all good, typer does the string -> `int` conversion and your code takes it from there.

File: ancv/web/server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexpovel) [reply ancv/web/server.py]: Yeah just readability. `200` is kind of a wash, could also go with the hard-coded number. For other codes I think the named constants are more readable, that's all!

File: ancv/web/server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexpovel) [reply ancv/web/server.py]: Ok sounds good! Tests will show and manifest that it behaves well.

File: ancv/web/server.py (hunk 4 above)
Comment (kiraum) [reply ancv/web/server.py]: Using self.cache as strong typed `str` in the function split refactoring, because I'm caching the template rendered from `Template.render()` which returns a `str`, done.

**Comment by alexpovel** [comment]
> Thanks for the quick answer/comments, I will go over it during the week, a bit slow here, because it's herfstvakantie (autumn holidays), so a bit busy with family! 😅

Sounds great! Let's take all the time we like, this is about having fun!

> `[project.optional-dependencies]`

I think I prefer `uv`'s approach. They give a [good reason](https://docs.astral.sh/uv/concepts/dependencies/#development-dependencies):

> Unlike optional dependencies, development dependencies are local-only and will not be included in the project requirements when published to PyPI or other indexes. As such, development dependencies are not included in the `[project]` table.

which is what we want. Those dependencies shouldn't occur anywhere but for local dev, and in CI. With optional deps, users could [`pip install ancv[dev]`](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#dependencies-optional-dependencies) if I read that guide right, which doesn't seem right.

**Comment by alexpovel** [comment]
Just discovered: I refactored to use `uv` just days ago, and used

https://github.com/alexpovel/ancv/blob/e60f5f55b3356428c3e59a916f9825a80ace0c64/pyproject.toml#L53-L67

which is already outdated practice...:

https://github.com/astral-sh/uv/releases/tag/0.4.27

aka PEP 735 will allow this to be expressed natively and standards-based, which is fantastic:

```toml
[dependency-groups]
dev = [
  "pytest >=8.1.1,<9"
]
```

etc.

But `0.4.27` isn't packaged for devbox/nix yet, so can't and don't want to rely on that just yet.

File: ancv/web/server.py (hunk 4 above)
Comment (kiraum) [reply ancv/web/server.py]: I made the `self.cache` strong typed `str`, since the implementation is dealing with the rendered template, do you think that's better to change it to `self.cache: Optional[ResumeSchema] = None`, and the functions to deal with that?

File: ancv/web/server.py [hunk 11]
```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        stopwatch(segment="Fetch Resume")
+
+        try:
+            async with session.get(self.destination) as response:
+                if response.status != 200:
+                    raise web.HTTPBadRequest(
+                        reason=f"Failed to fetch resume from {self.destination}"
+                    )
+                content = await response.text()
+                resume_json = json.loads(content)
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Done.

**Comment by kiraum** [comment]
> Just discovered: I refactored to use `uv` just days ago, and used
> 
> https://github.com/alexpovel/ancv/blob/e60f5f55b3356428c3e59a916f9825a80ace0c64/pyproject.toml#L53-L67
> 
> which is already outdated practice...:
> 
> https://github.com/astral-sh/uv/releases/tag/0.4.27
> 
> aka PEP 735 will allow this to be expressed natively and standards-based, which is fantastic:
> 
> ```toml
> [dependency-groups]
> dev = [
>   "pytest >=8.1.1,<9"
> ]
> ```
> 
> etc.
> 
> But `0.4.27` isn't packaged for devbox/nix yet, so can't and don't want to rely on that just yet.

yeah, uv release is too fast, hard to follow 😅 

I will give a try to that myself this week!

File: ancv/web/server.py (hunk 4 above)
Comment (alexpovel) [reply ancv/web/server.py]: Hmmm, up to you!

In my mind the cache is for network I/O, and going from `ResumeSchema` to a rendered `str` is so cheap and fast it won't matter unless you're requesting your resume at crazy RPS. I just really hate `str` as a type. It says and guarantees basically nothing. But if `cache` as `str` is a better fit, it's okay!

File: ancv/web/server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Removed the logic using tmp file in favor of ResumeSchema, and caching in memory. Done.

**Comment by kiraum** [comment]
@alexpovel , sorry for the delay, busy days here... company announced reorg, so a lot happening! 😅

Added a section about the new function in the README.md, and tests to verify that WebHandler correctly fetches, caches and serves resume data from a URL while handling both successful and error scenarios.

I checked ruff check/format, mypy, and pytest, results looks good, can you please take a look when possible? If good, can  run the checks again?

File: tests/web/test_server.py [hunk 12]
```diff
@@ -299,3 +301,72 @@ def test_server_timing_header(
...
+
+        # Test initial fetch
+        resp = await client.get("/")
+        assert resp.status == HTTPStatus.OK
+        content = await resp.text()
+        assert "Test User" in content
+        assert "Developer" in content
+
+        # Test caching
+        first_response = content
```
Comment (alexpovel) [inline tests/web/test_server.py:341]: This doesn't _really_ test caching. It might just fetch without caching again, no?

You could, additionally to what you're doing here, further up do:

```python
        hitcount = 0
        # Set up a mock resume server
        async def mock_resume_handler(request):
            nonlocal hitcount
            hitcount += 1
            return Response(
                text=json.dumps({"basics": {"name": "Test User", "label": "Developer"}})
            )
```

(paraphrasing syntax here, might not be right)

then check that `hitcount` is still only 1 after this hopefully cached fetch.

File: tests/web/test_server.py [hunk 13]
```diff
@@ -299,3 +301,72 @@ def test_server_timing_header(
...
+        resp = await client.get("/")
+        assert await resp.text() == first_response
+
+        # Test refresh after interval
+        await asyncio.sleep(refresh_interval.total_seconds() + 0.1)
+        resp = await client.get("/")
+        assert resp.status == HTTPStatus.OK
+        assert await resp.text() == first_response  # Should be same content
+
```
Comment (alexpovel) [inline tests/web/test_server.py:350]: and then here you can add `assert hitcount == 2`

File: tests/web/test_server.py [hunk 14]
```diff
@@ -299,3 +301,72 @@ def test_server_timing_header(
...
+class TestWebHandler:
+    async def test_web_handler_basic_functionality(
+        self,
+        aiohttp_client: Any,
+        aiohttp_server: Any,
+    ) -> None:
+        # Set up a mock resume server
+        async def mock_resume_handler(request):
+            return Response(
+                text=json.dumps({"basics": {"name": "Test User", "label": "Developer"}})
```
Comment (alexpovel) [inline tests/web/test_server.py:317]: You could use: https://docs.aiohttp.org/en/stable/_modules/aiohttp/web_response.html#json_response

File: ancv/web/server.py [hunk 15]
```diff
@@ -258,3 +262,174 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+
+        Args:
+            session: The aiohttp client session to use for requests.
+
+        Returns:
+            ResumeSchema: The validated resume data
+
+        Raises:
+            ResumeLookupError: When resume cannot be fetched from destination
+            json.JSONDecodeError: When response is not valid JSON
```
Comment (alexpovel) [inline ancv/web/server.py:329]: This one (and I think others) isn't actually thrown, is it? It's swallowed, and callers won't need to/can't handle it

File: ancv/web/server.py [hunk 16]
```diff
@@ -258,3 +262,174 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
...
+        )
+
+        if should_refresh:
+            log.debug("Fetching fresh resume data.")
+            try:
+                resume_data = await self.fetch(session)
+                if isinstance(resume_data, web.Response):
+                    return resume_data
+                rendered = self.render(resume_data)
+                if isinstance(rendered, web.Response):
```
Comment (alexpovel) [inline ancv/web/server.py:410]: Let's get rid of all `isinstance` checks! What this effectively does is **mix** error handling: now we have a `typing.Union` of `str | Response`, **and** we have some normal Python exceptions as well, and have to handle both.

I think a better design is for `render` to only return `str` and raise in _all_ other cases. While exceptions are really annoying to work with, it's more in line with "normal Python" and importantly it's only one error mechanism.

For that to work, you could make your own error

```python
class SomeException(Exception):
    """Raised when ..."""
```

and `raise` that manually. You could also inherit from `ValueError` or whatever if you think that's more suitable (don't think it's necessary).

And finally, you could give it a payload error (OR create multiple new exception types, each without a payload... both should work). Then, in this top-level `root` function, we just do:

```python
try:
    x = y()
except this:
   pass
except that:
   pass
```

Like you're already doing, just more of it 🙂 A final result shouldn't need `isinstance` checks.

**Review by alexpovel — COMMENTED** [review]
Thanks, no worries!

It's looking good, thanks for adding tests and docs. I just have comments on some of the error handling (and tests thereof). Then it's good to go!

File: ancv/web/server.py [hunk 17]
```diff
@@ -258,3 +262,174 @@
...
+            ValueError: When template rendering fails
+        """
+        try:
+            template = Template.from_model_config(resume_data)
+            rendered = template.render()
+            if isinstance(rendered, str):
+                return rendered
+            return web.Response(text="Template rendering failed")
+        except ResumeConfigError as exc:
+            return web.Response(text=str(exc))
```
[the remark this answers is not in the document]
Comment (alexpovel) [reply ancv/web/server.py:371]: Yeah fair point, you can just return a generic `web.Response(text="Resume configuration error", status=HTTPStatus.INTERNAL_SERVER_ERROR)` or similar

**Comment by alexpovel** [comment]
https://github.com/alexpovel/ancv/actions/runs/11845227596/job/33079082336?pr=241#step:7:1315

These errors are on me, sorry ☹️ Thought I fixed those. It's something about PRs and having access to secrets... I'm happy to test your PR locally and then merge anyway

File: tests/web/test_server.py (hunk 12 above)
Comment (kiraum) [reply tests/web/test_server.py:345]: I like the suggestion, update trying to achieve it, can you please double check? I put all the changes on a diff [commit](https://github.com/alexpovel/ancv/pull/241/commits/f87d2a013e3a621c99b8f7ce28135a0cd4b3988b) to make it easier to check

File: tests/web/test_server.py (hunk 13 above)
Comment (kiraum) [reply tests/web/test_server.py:356]: Yep, following the answer above.

File: tests/web/test_server.py (hunk 14 above)
Comment (kiraum) [reply tests/web/test_server.py]: Done following the suggestion, can you please take a look?

File: ancv/web/server.py (hunk 15 above)
Comment (kiraum) [reply ancv/web/server.py]: Docstrings updated, please let me know if you notice something wrong. That was my bad, wrote in the beginning, and after few changes, they got outdated/wrong, I did reviewed all, and they seems correct now.

File: ancv/web/server.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]: Agree, changed logic here, can you please double check?

**Comment by kiraum** [comment]
No problems, and don't worry, really appreciate the comments/time to review it, and I always do basic tests locally before to push, so they should be good..... in theory 😅

**Comment by kiraum** [comment]
My bad, just pushed it, should be good now! =]

File: ancv/web/server.py [hunk 18]
```diff
@@ -258,3 +262,170 @@
...
+                    return web.Response(
+                        text="No cache available", status=HTTPStatus.SERVICE_UNAVAILABLE
+                    )
+            except (RenderError, InvalidResumeDataError) as exc:
+                log.error("Resume rendering error", error=str(exc))
+                if self._last_valid_render:
+                    self.cache = self._last_valid_render
+                    log.warning("Using last valid render as fallback")
+                else:
+                    return web.Response(text=str(exc), status=HTTPStatus.BAD_REQUEST)
```
[the remark this answers is not in the document]
Comment (kiraum) [reply ancv/web/server.py]:
```
-                    return web.Response(text=str(exc), status=HTTPStatus.BAD_REQUEST)
+                    return web.Response(
+                        text="Unable to render resume",
+                        status=HTTPStatus.INTERNAL_SERVER_ERROR,
+                    )
```

**Comment by alexpovel** [comment]
Thank you for your contribution @kiraum ! I just merged... let's see what fails in CI now. I need to fix that flakiness...

**Comment by kiraum** [comment]
Thanks! My pleasure, learning a lot here! =]
